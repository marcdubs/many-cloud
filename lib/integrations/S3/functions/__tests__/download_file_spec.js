describe("download_file", () => {
  describe("when the file is downloaded successfully", () => {
    const expected_dest = "some_file_dest.extension",
      expected_file_id = "1234",
      expected_output = "some_file_data";
    let func, data, stream, fs, result, createReadStream;
    beforeEach(async done => {
      fs = require("fs");
      fs.createWriteStream = jest.fn().mockImplementation(() => {
        return expected_output;
      });
      stream = {
        pipe: jest.fn()
      };

      createReadStream = jest.fn().mockImplementation(() => {
        return stream;
      });

      data = {
        s3: {
          getObject: jest.fn().mockImplementation(() => {
            return { createReadStream: createReadStream };
          })
        }
      };
      func = require("../download_file")(data);
      result = await func(expected_file_id, expected_dest);
      done();
    });

    it("calls fs.createWriteStream with dest", () => {
      expect(fs.createWriteStream).toHaveBeenCalledWith(expected_dest);
    });

    it("calls stream.pipe with what createWriteStream returned", () => {
      expect(stream.pipe).toHaveBeenCalledWith(expected_output);
    });

    it("returns undefined", () => {
      expect(result).toBeUndefined();
    });
  });

  describe("when it returns an error", () => {
    const expected_error =
      "It's a lot of bad things That they wishin' and wishin' and wishin' and wishin' They wishin' on me Yuh, ayy, ayy";
    beforeEach(() => {
      data = {
        s3: {
          getObject: jest.fn().mockImplementation(() => {
            return {
              createReadStream: jest.fn().mockImplementation(() => {
                return {
                  pipe: jest.fn().mockImplementation(() => {
                    throw new Error(expected_error);
                  })
                };
              })
            };
          })
        }
      };
      func = require("../download_file")(data);
    });

    it("rejects with the error", async () => {
      await expect(func()).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
