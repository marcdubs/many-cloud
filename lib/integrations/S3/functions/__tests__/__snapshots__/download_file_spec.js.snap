// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`download_file when it returns an error rejects with the error 1`] = `"It's a lot of bad things That they wishin' and wishin' and wishin' and wishin' They wishin' on me Yuh, ayy, ayy"`;
